### PR TITLE
[fix #700][fix #790] open file in bookshelf without re-import it

### LINKED ISSUE #700: same book imported multiple times into local bookshelf, we need to discriminate the unicity of epub file (hash ?)
Tested with `npm run cli -- PATH_TO_EPUB`

Related issue: https://github.com/readium/readium-desktop/issues/438

### LINKED ISSUE #790: no error message in frontend when epub import has failed, only on import with cli (open with)
command to reduce the epub size : 
`dd if=Rene+Descartes+-+Discours+de+la+methode.epub of=Rene+Descartes+-+Discours+de+la+methode.min.epub bs=1k count=100`

[Rene+Descartes+-+Discours+de+la+methode.epub.zip](https://github.com/readium/readium-desktop/files/3717885/Rene%2BDescartes%2B-%2BDiscours%2Bde%2Bla%2Bmethode.epub.zip)

### PR DESCRIPTION
**Pull request by panaC**
fix #700 
fix #790 

To generate a unique Hash/Checksum on zip file (epub/epub3/cbz/audiobook/...). 
- I parse the zip file with a zip node library ([zip file format](https://en.wikipedia.org/wiki/ZIP_(file_format)))
- I get all crc32 in each header file in zip
- I create a unique sha1 from crc32 array.

This method is very quickly, just the calculation time to parse zip file and the generation of a new hash (data: 4Bytes * N files in zip).

### CODE REVIEW

**Comment by danielweck** [comment]
I have some test results and benchmark data to contribute.

The `r2-utils-js` package currently references 3 different zip libraries: `node-stream-zip`, `yauzl` and `unzipper`.
The `r2-utils-js` codebase contains a benchmark script which has been used to compare performance metrics when opening / streaming EPUB resources from large zip archives.

I have added a new benchmark script that scans the ZIP directory of EPUB publications and extracts pre-computed CRC(32) checksums associated with each publication resource. This script verifies the order of directory entries, and the equality of their CRC numbers.

I am glad to report that all 3 ZIP libraries return the exact same CRC figures, in the exact same order (we just have to be careful with `yauzl` which returns zero CRCs, whereas `unzipper` and `node-stream-zip` do not).

The `node-stream-zip` library is consistently (without exceptions) the fastest zip directory processor. Note that this does not necessarily mean that it is also the fastest when streaming / piping resource bytes (this would have to be tested separately).

The `yauzl` library is second best with smaller publications, whereas `unzipper` takes the lead with larger EPUB files (see some test results below).

You may try the benchmark script yourself:
1) `cd SOMEWHERE`
2) `git clone https://github.com/readium/r2-utils-js.git`
3) `cd r2-utils-js`
4) `npm install`
5) `npm run build`
6) `npm run cli-crc "PATH_TO_EPUB"` or `npm run cli-src "PATH_TO_FOLDER"` (paths can be absolute or relative)

TypeScript source code:
https://github.com/readium/r2-utils-js/blob/develop/src/_utils/perf-zip-crc-cli.ts

Some results:

```
// medea.epub
// >> Zip 1 (node-stream-zip) => 2,992,624 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 12,762,112 nanoseconds [ +9,769,488 ]
// >> Zip 3 (unzipper) => 7,472,482 nanoseconds [ +4,479,858 ]

// CotCotCot_MaMamieEnPoevie.epub
// >> Zip 1 (node-stream-zip) => 3,402,988 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 15,019,870 nanoseconds [ +11,616,882 ]
// >> Zip 3 (unzipper) => 9,025,568 nanoseconds [ +5,622,580 ]

// Bikes+Boards22.epub
// >> Zip 1 (node-stream-zip) => 526,562 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,962,268 nanoseconds [ +1,435,706 ]
// >> Zip 3 (unzipper) => 1,869,783 nanoseconds [ +1,343,221 ]

// BarrierefreiePdfDokumente.epub
// >> Zip 1 (node-stream-zip) => 2,081,902 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 8,240,815 nanoseconds [ +6,158,913 ]
// >> Zip 3 (unzipper) => 5,093,388 nanoseconds [ +3,011,486 ]

// cc-shared-culture.epub
// >> Zip 1 (node-stream-zip) => 249,000 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 778,500 nanoseconds [ +529,500 ]
// >> Zip 3 (unzipper) => 984,467 nanoseconds [ +735,467 ]

// GG-1.0.5_SAMPLE.epub
// >> Zip 1 (node-stream-zip) => 1,852,917 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 7,450,482 nanoseconds [ +5,597,565 ]
// >> Zip 3 (unzipper) => 4,560,785 nanoseconds [ +2,707,868 ]

// Heroes_Stephen_Fry.epub
// >> Zip 1 (node-stream-zip) => 497,732 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,990,731 nanoseconds [ +1,492,999 ]
// >> Zip 3 (unzipper) => 1,638,490 nanoseconds [ +1,140,758 ]

// white-fang.epub
// >> Zip 1 (node-stream-zip) => 557,335 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 2,145,522 nanoseconds [ +1,588,187 ]
// >> Zip 3 (unzipper) => 1,914,046 nanoseconds [ +1,356,711 ]

// kusamakura-japanese-vertical-writing.epub
// >> Zip 1 (node-stream-zip) => 272,317 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 834,772 nanoseconds [ +562,455 ]
// >> Zip 3 (unzipper) => 1,037,936 nanoseconds [ +765,619 ]

// Anna's_Day_of_Gratitude_ePUB3_Audio.epub
// >> Zip 1 (node-stream-zip) => 507,889 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,891,184 nanoseconds [ +1,383,295 ]
// >> Zip 3 (unzipper) => 1,693,999 nanoseconds [ +1,186,110 ]

// llamalla_cbuscepgi034l30h.epub
// >> Zip 1 (node-stream-zip) => 1,652,412 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 6,692,874 nanoseconds [ +5,040,462 ]
// >> Zip 3 (unzipper) => 8,131,035 nanoseconds [ +6,478,623 ]

// KM517_CallOfTheWild_MultiGranularityEPUB3MO.epub
// >> Zip 1 (node-stream-zip) => 348,856 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,101,180 nanoseconds [ +752,324 ]
// >> Zip 3 (unzipper) => 1,669,794 nanoseconds [ +1,320,938 ]

// GiveMyRegards-to-BlackJack1.epub
// >> Zip 1 (node-stream-zip) => 1,980,867 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 8,178,036 nanoseconds [ +6,197,169 ]
// >> Zip 3 (unzipper) => 4,959,790 nanoseconds [ +2,978,923 ]

// Enseigner les maths au cycle 4_EPUB.v1.0.epub
// >> Zip 1 (node-stream-zip) => 1,044,097 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 4,279,604 nanoseconds [ +3,235,507 ]
// >> Zip 3 (unzipper) => 3,015,442 nanoseconds [ +1,971,345 ]

// mahabharata.epub
// >> Zip 1 (node-stream-zip) => 8,521,493 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 37,148,213 nanoseconds [ +28,626,720 ]
// >> Zip 3 (unzipper) => 22,458,262 nanoseconds [ +13,936,769 ]

// 80968603-epub3.epub
// >> Zip 1 (node-stream-zip) => 671,227 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 2,559,357 nanoseconds [ +1,888,130 ]
// >> Zip 3 (unzipper) => 2,118,335 nanoseconds [ +1,447,108 ]

// jlreq-in-japanese.epub
// >> Zip 1 (node-stream-zip) => 2,265,684 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 9,475,816 nanoseconds [ +7,210,132 ]
// >> Zip 3 (unzipper) => 5,682,153 nanoseconds [ +3,416,469 ]

// MB_BiB_2014B.epub
// >> Zip 1 (node-stream-zip) => 375,677 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,442,590 nanoseconds [ +1,066,913 ]
// >> Zip 3 (unzipper) => 1,623,231 nanoseconds [ +1,247,554 ]

// Advanced-Accessibility-Tests-Media-Overlays-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 214,448 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 518,818 nanoseconds [ +304,370 ]
// >> Zip 3 (unzipper) => 944,622 nanoseconds [ +730,174 ]

// epub30-test-0320-20160325.epub
// >> Zip 1 (node-stream-zip) => 229,417 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 540,365 nanoseconds [ +310,948 ]
// >> Zip 3 (unzipper) => 866,387 nanoseconds [ +636,970 ]

// SpacePirates_MultiGranularityEPUB3MO.epub
// >> Zip 1 (node-stream-zip) => 224,915 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 504,224 nanoseconds [ +279,309 ]
// >> Zip 3 (unzipper) => 922,994 nanoseconds [ +698,079 ]

// alice3.epub
// >> Zip 1 (node-stream-zip) => 815,715 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 3,242,681 nanoseconds [ +2,426,966 ]
// >> Zip 3 (unzipper) => 2,350,496 nanoseconds [ +1,534,781 ]

// ladybugg_cbuscepgi034l00e.epub
// >> Zip 1 (node-stream-zip) => 601,518 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,715,584 nanoseconds [ +1,114,066 ]
// >> Zip 3 (unzipper) => 2,500,082 nanoseconds [ +1,898,564 ]

// Accessible-EPUB-3.epub
// >> Zip 1 (node-stream-zip) => 301,979 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 946,513 nanoseconds [ +644,534 ]
// >> Zip 3 (unzipper) => 1,282,277 nanoseconds [ +980,298 ]

// douglas-kennedy-the-pickup.epub
// >> Zip 1 (node-stream-zip) => 350,922 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,127,616 nanoseconds [ +776,694 ]
// >> Zip 3 (unzipper) => 1,411,462 nanoseconds [ +1,060,540 ]

// bibliotherapie-lire-c-est-guerir.epub
// >> Zip 1 (node-stream-zip) => 458,744 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,992,308 nanoseconds [ +1,533,564 ]
// >> Zip 3 (unzipper) => 1,765,000 nanoseconds [ +1,306,256 ]

// page-blanche.epub
// >> Zip 1 (node-stream-zip) => 267,045 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 792,148 nanoseconds [ +525,103 ]
// >> Zip 3 (unzipper) => 1,035,116 nanoseconds [ +768,071 ]

// anton-chekhov_short-fiction.epub3
// >> Zip 1 (node-stream-zip) => 1,059,635 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 4,058,142 nanoseconds [ +2,998,507 ]
// >> Zip 3 (unzipper) => 2,981,690 nanoseconds [ +1,922,055 ]

// BellaOriginal3.epub
// >> Zip 1 (node-stream-zip) => 560,176 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,677,144 nanoseconds [ +1,116,968 ]
// >> Zip 3 (unzipper) => 1,759,150 nanoseconds [ +1,198,974 ]

// Dumas - Le Comte de Monte Cristo-37.epub
// >> Zip 1 (node-stream-zip) => 767,417 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 2,881,553 nanoseconds [ +2,114,136 ]
// >> Zip 3 (unzipper) => 3,590,771 nanoseconds [ +2,823,354 ]

// miguel-de-cervantes-saavedra_don-quixote.epub3
// >> Zip 1 (node-stream-zip) => 843,643 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 2,939,217 nanoseconds [ +2,095,574 ]
// >> Zip 3 (unzipper) => 2,325,187 nanoseconds [ +1,481,544 ]

// moby-dick.epub
// >> Zip 1 (node-stream-zip) => 842,785 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 3,095,438 nanoseconds [ +2,252,653 ]
// >> Zip 3 (unzipper) => 2,309,067 nanoseconds [ +1,466,282 ]

// israelsailing.epub
// >> Zip 1 (node-stream-zip) => 320,358 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 839,174 nanoseconds [ +518,816 ]
// >> Zip 3 (unzipper) => 1,013,928 nanoseconds [ +693,570 ]

// Contes_fantastiques_(3e_éd_)_[...]Erckmann-Chatrian_Auteur_bpt6k65669461.epub
// >> Zip 1 (node-stream-zip) => 380,339 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 912,758 nanoseconds [ +532,419 ]
// >> Zip 3 (unzipper) => 1,170,077 nanoseconds [ +789,738 ]

// moonstone_Wilkie+Collins+-+The+Moonstone.epub
// >> Zip 1 (node-stream-zip) => 569,036 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,941,845 nanoseconds [ +1,372,809 ]
// >> Zip 3 (unzipper) => 2,700,147 nanoseconds [ +2,131,111 ]

// The-Art-of-War.epub
// >> Zip 1 (node-stream-zip) => 596,495 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,517,918 nanoseconds [ +921,423 ]
// >> Zip 3 (unzipper) => 1,627,116 nanoseconds [ +1,030,621 ]

// linear-algebra.epub
// >> Zip 1 (node-stream-zip) => 669,014 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 2,075,956 nanoseconds [ +1,406,942 ]
// >> Zip 3 (unzipper) => 1,953,857 nanoseconds [ +1,284,843 ]

// cole-voyage-of-life.epub
// >> Zip 1 (node-stream-zip) => 323,875 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 723,803 nanoseconds [ +399,928 ]
// >> Zip 3 (unzipper) => 963,742 nanoseconds [ +639,867 ]

// E-book Watermarking Whitepaper.epub
// >> Zip 1 (node-stream-zip) => 283,529 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 615,583 nanoseconds [ +332,054 ]
// >> Zip 3 (unzipper) => 934,687 nanoseconds [ +651,158 ]

// wasteland-otf-obf.epub
// >> Zip 1 (node-stream-zip) => 201,793 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 491,837 nanoseconds [ +290,044 ]
// >> Zip 3 (unzipper) => 972,474 nanoseconds [ +770,681 ]

// moonstone_pg155-images.epub
// >> Zip 1 (node-stream-zip) => 555,294 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,413,821 nanoseconds [ +858,527 ]
// >> Zip 3 (unzipper) => 1,590,957 nanoseconds [ +1,035,663 ]

// moonstone.epub
// >> Zip 1 (node-stream-zip) => 560,323 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,487,910 nanoseconds [ +927,587 ]
// >> Zip 3 (unzipper) => 2,085,212 nanoseconds [ +1,524,889 ]

// michel_zevaco-les_amours_du_chico.epub
// >> Zip 1 (node-stream-zip) => 423,942 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 1,100,478 nanoseconds [ +676,536 ]
// >> Zip 3 (unzipper) => 1,286,217 nanoseconds [ +862,275 ]

// georgia-pls-ssml.epub
// >> Zip 1 (node-stream-zip) => 211,471 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 400,459 nanoseconds [ +188,988 ]
// >> Zip 3 (unzipper) => 977,445 nanoseconds [ +765,974 ]

// Liza+Palmer+-+The+Nobodies_preview.epub
// >> Zip 1 (node-stream-zip) => 259,864 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 573,368 nanoseconds [ +313,504 ]
// >> Zip 3 (unzipper) => 882,962 nanoseconds [ +623,098 ]

// arabic.epub
// >> Zip 1 (node-stream-zip) => 242,495 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 657,206 nanoseconds [ +414,711 ]
// >> Zip 3 (unzipper) => 952,193 nanoseconds [ +709,698 ]

// alice_Lewis+Carroll+-+Alices+Adventures+in+Wonderland.epub
// >> Zip 1 (node-stream-zip) => 355,847 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 886,048 nanoseconds [ +530,201 ]
// >> Zip 3 (unzipper) => 1,764,819 nanoseconds [ +1,408,972 ]

// HolmesMixture.epub
// >> Zip 1 (node-stream-zip) => 367,724 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 902,662 nanoseconds [ +534,938 ]
// >> Zip 3 (unzipper) => 1,218,852 nanoseconds [ +851,128 ]

// Zenaro esercitazione ebook.epub
// >> Zip 1 (node-stream-zip) => 265,509 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 612,123 nanoseconds [ +346,614 ]
// >> Zip 3 (unzipper) => 879,128 nanoseconds [ +613,619 ]

// ArieBaalGoof-epub2.epub
// >> Zip 1 (node-stream-zip) => 342,435 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 874,741 nanoseconds [ +532,306 ]
// >> Zip 3 (unzipper) => 1,037,003 nanoseconds [ +694,568 ]

// mymedia_lite.epub
// >> Zip 1 (node-stream-zip) => 303,743 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 747,752 nanoseconds [ +444,009 ]
// >> Zip 3 (unzipper) => 1,112,899 nanoseconds [ +809,156 ]

// marie-de-sade-mein-facettenreich.epub
// >> Zip 1 (node-stream-zip) => 347,854 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 914,372 nanoseconds [ +566,518 ]
// >> Zip 3 (unzipper) => 1,201,887 nanoseconds [ +854,033 ]

// Fundamental-Accessibility-Tests-Non-Visual-Reading-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 227,626 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 488,372 nanoseconds [ +260,746 ]
// >> Zip 3 (unzipper) => 884,371 nanoseconds [ +656,745 ]

// Advanced-Accessibility-Tests-Extended-Descriptions-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 312,512 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 620,444 nanoseconds [ +307,932 ]
// >> Zip 3 (unzipper) => 924,929 nanoseconds [ +612,417 ]

// childrens-literature.epub
// >> Zip 1 (node-stream-zip) => 197,975 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 365,502 nanoseconds [ +167,527 ]
// >> Zip 3 (unzipper) => 903,286 nanoseconds [ +705,311 ]

// Fundamental-Accessibility-Tests-Visual-Adjustments-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 205,454 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 472,168 nanoseconds [ +266,714 ]
// >> Zip 3 (unzipper) => 906,123 nanoseconds [ +700,669 ]

// hsmmt10t.epub
// >> Zip 1 (node-stream-zip) => 346,137 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 906,219 nanoseconds [ +560,082 ]
// >> Zip 3 (unzipper) => 1,093,459 nanoseconds [ +747,322 ]

// hebrew.epub
// >> Zip 1 (node-stream-zip) => 205,796 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 470,673 nanoseconds [ +264,877 ]
// >> Zip 3 (unzipper) => 916,742 nanoseconds [ +710,946 ]

// Advanced-Accessibility-Tests-Mathematics-v1.1.0.epub
// >> Zip 1 (node-stream-zip) => 281,812 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 594,602 nanoseconds [ +312,790 ]
// >> Zip 3 (unzipper) => 921,549 nanoseconds [ +639,737 ]

// regime-anticancer-arabic.epub
// >> Zip 1 (node-stream-zip) => 210,703 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 418,531 nanoseconds [ +207,828 ]
// >> Zip 3 (unzipper) => 945,621 nanoseconds [ +734,918 ]

// Fundamental-Accessibility-Tests-Basic-Functionality-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 205,123 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 415,854 nanoseconds [ +210,731 ]
// >> Zip 3 (unzipper) => 930,226 nanoseconds [ +725,103 ]

// Fundamental-Accessibility-Tests-Read-Aloud-v1.0.0.epub
// >> Zip 1 (node-stream-zip) => 202,085 nanoseconds  [ WINNER ]
// >> Zip 2 (yauzl) => 379,464 nanoseconds [ +177,379 ]
// >> Zip 3 (unzipper) => 934,516 nanoseconds [ +732,431 ]
```

File: src/main/crc.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                reject(err);
+            }
+            data.on("error", (e) => reject(e));
+            data.readEntry();
+            data.on("entry", (entry: yauzl.Entry) => {
+                fileArray.push([entry.fileName, (entry.crc32 || 0).toString(16)]);
+                data.readEntry();
+            });
+            data.on("end", () => {
+                const sortArray = fileArray.sort((a, b) => a[0] === b[0] ? 0 : a[0] < b[0] ? -1 : 1);
```
Comment (danielweck) [inline src/main/crc.ts]: Based on my tests with a large EPUB set, sorting is not necessary.
See https://github.com/readium/readium-desktop/pull/793#issuecomment-542363462

File: src/main/crc.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    return new Promise<string>((resolve, reject) => {
+
+        yauzl.open(filePath, { lazyEntries: true }, (err, data) => {
+            if (err) {
+                reject(err);
+            }
+            data.on("error", (e) => reject(e));
+            data.readEntry();
+            data.on("entry", (entry: yauzl.Entry) => {
+                fileArray.push([entry.fileName, (entry.crc32 || 0).toString(16)]);
```
Comment (danielweck) [inline src/main/crc.ts]: Based on my test, with some EPUBs `yauzl` returns `0` CRC for some ZIP directory entries.
Therefore I would suggest filtering out zero values (e.g. `arr.filter((val) => { return val; })`), in order to be consistent with other ZIP libs such as `unzipper` and `node-stream-zip`.

See https://github.com/readium/readium-desktop/pull/793#issuecomment-542363462

File: src/main/crc.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import { crc32 } from "crc";
+import * as yauzl from "yauzl";
+
+type TCrcFile = [string, string];
```
Comment (danielweck) [inline src/main/crc.ts]: Just a thought: in my tests, the CRC numbers are quite short and human-readable. I would therefore prefer to define `type TCrcFile = [string, number]`, and I would postpone the use of `.toString(16)` to the last minute inside the `end` event handler (i.e. `crc32(curr[1].toString(16), prev)`), instead of the current code `fileArray.push([entry.fileName, (entry.crc32 || 0).toString(16)]);` inside the `entry` event handler.

This way, the `entry` event is super simple (collects raw data), and the `end` event contains more logic to process the collected data.

File: src/main/crc.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+export async function extractCrc32OnZip(filePath: string) {
+    const fileArray: TCrcFile[] = [];
+    return new Promise<string>((resolve, reject) => {
+
+        yauzl.open(filePath, { lazyEntries: true }, (err, data) => {
+            if (err) {
+                reject(err);
```
Comment (danielweck) [inline src/main/crc.ts]: Missing return statement (early closure block exit).

File: src/main/crc.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            }
+            data.on("error", (e) => reject(e));
+            data.readEntry();
+            data.on("entry", (entry: yauzl.Entry) => {
```
Comment (danielweck) [inline src/main/crc.ts]: Note that `entry` can be a folder (no CRC), so should preferably be skipped with  `entry.fileName[entry.fileName.length - 1] === "/"` (because `entry.crc32` can actually be the number `0`, not just `null` or `undefined`)

**Comment by danielweck** [comment]
Note that I am able to reproduce a bug with `unzipper`, so probably better stick with either `yauzl` or `node-zip-stream`. See:
https://github.com/ZJONSSON/node-unzipper/issues/104#issuecomment-542697289

File: src/main/crc.ts (hunk 1 above)
Comment (panaC) [reply src/main/crc.ts]: Based on my tests on Thorium, I think is necessary to sorting. We cannot be sure that the order received on entry event is correct

File: src/main/crc.ts (hunk 3 above)
Comment (panaC) [reply src/main/crc.ts]: fix 84970dd8

File: src/main/crc.ts (hunk 4 above)
Comment (panaC) [reply src/main/crc.ts]: fix 84970dd8

File: src/main/crc.ts (hunk 5 above)
Comment (panaC) [reply src/main/crc.ts]: I prefer to take the directories, they will be counted in the checksum. If a empty directory is added in zip the checksum will be different

What do you think ?

**Comment by panaC** [comment]
I think yauzl is still a good deal. the mean size of a publication is about 1MB, yauzl still a best candidate.

What do you think ? I let my code with yauzl lib ?

**Comment by panaC** [comment]
I replaced crc32 with sha1, the collision risk rise up if lot of publication added to thorium.

**Comment by panaC** [comment]
Quid of the opds catalog publication, currently a message is asked to user if he want re-import an opds publication : We can lets the user take the decision or download each time the opds publication and check if already added in back. (with the hash) ?

File: src/main/crc.ts (hunk 1 above)
Comment (danielweck) [reply src/main/crc.ts]: I have extensively batch-tested (and stress-tested) this assumption in the `r2-utils-js` using an automated benchmarking script, so I beg to differ :)
Do you have data to backup this claim, or just anecdotal evidence?

File: src/main/crc.ts (hunk 2 above)
Comment (danielweck) [reply src/main/crc.ts]: Good point!

File: src/main/crc.ts (hunk 5 above)
Comment (danielweck) [reply src/main/crc.ts]: Good point!

**Comment by danielweck** [comment]
`yauzl` is best overall, based on the benchmark results (overall ZIP directory access and ZIP streaming)

**Comment by danielweck** [comment]
Regarding OPDS: I think that the originating URL of a given publication is its canonical "unique identifier" (with respect to identifying duplicates in the existing library) ... assuming the offline database preserves the originating URL of a publication acquired via OPDS (is that the case, or are we currently using a different heuristic, like comparing `dc:identifier`?)

But, there is a possible scenario of downloading a publication from an OPDS feed using an external web browser (or some other client), and to subsequently side-load the publication into Thorium (let's call this `PUBLICATION_1`). Then, if the end-user attempts to import the exact same publication from the OPDS feed using Thorium's catalog browser (`PUBLICATION_2`), there isn't a collision due to matching originating URLs (because `PUBLICATION_1` was imported from the filesystem), so Thorium will download the EPUB before being able to compare the CRC+SHA1 unique hash.
In other words, the true test is always to compute incoming publication's hash, then compare with the library database.

File: src/main/crc.ts (hunk 1 above)
Comment (danielweck) [reply src/main/crc.ts]: Sorting is guaranteed with `yauzl` despite the asynchronous `entry` event, thanks to the `readEntry()` function call. So we can eliminate the string sorting :)

File: src/main/api/publication.ts [hunk 6]
```diff
@@ -179,40 +178,22 @@ export class PublicationApi implements IPublicationApi {
                 this.dispatchToastRequest(ToastType.DownloadFailed, `[${error}]`);
                 throw new Error(`importOpdsPublication ${error}`);
             }
-            titleView = publication.title;
             this.sendDownloadSuccess(url);
             returnView = this.publicationViewConverter.convertDocumentToView(publication);
         }
-
-        // dispatch notification to user with redux
-        this.dispatchToastRequest(ToastType.DownloadComplete,
```
Comment (danielweck) [inline src/main/api/publication.ts]: I see you've removed some `dispatchToastRequest`, but there is another one above ( `this.dispatchToastRequest(ToastType.DownloadFailed, `[${error}]`);`).
Is that an oversight, or is this intentional?

Comment (panaC) [reply src/main/api/publication.ts]: yes, it's intentional to catch only download error. It can be move to catalog service also. And handle the toast dispatch only on service

File: src/main/api/publication.ts [hunk 7]
```diff
@@ -179,40 +178,22 @@ export class PublicationApi implements IPublicationApi {
...
-
-        return newDocs.map((doc) => {
-            const publication = this.publicationViewConverter.convertDocumentToView(doc);
-            this.dispatchToastRequest(ToastType.DownloadComplete,
-                this.translator.translate("message.import.success", { title: publication.title }));
-            return publication;
-        });
+        // returns all publications linked to this import
+        const pubsRawPromise = filePathArray.map((filePath) => this.catalogService.importFile(filePath));
+        const pubsRaw = await Promise.all(pubsRawPromise);
```
Comment (danielweck) [inline src/main/api/publication.ts]: As soon as one import fails, the entire multi-file import fails.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

Should we allow individual imports to succeed, even if one of them fails?

File: src/main/cli/commandLine.ts [hunk 8]
```diff
@@ -6,16 +6,17 @@
...
 import { URL } from "url";
 import { isArray } from "util";
 
-export async function cli_(filePath: string) {
-    // import and read publication
-    const catalogService = diMainGet("catalog-service");
-    const publication = await catalogService.importFile(filePath);
-    const store = diMainGet("store");
+function openReader(publication: PublicationView | PublicationView[]) {
+    if (isArray(publication)) {
```
Comment (danielweck) [inline src/main/cli/commandLine.ts]: Should be negative (`!isArray()`)

File: src/renderer/components/dialog/publicationInfos/opdsControls.tsx [hunk 9]
```diff
@@ -62,4 +64,11 @@ const mapDispatchToProps = (dispatch: TDispatch) => {
     };
 };
 
-export default connect(undefined, mapDispatchToProps)(withTranslator(OpdsControls));
+// any because recursive type doesn't works
+const mapStateToProps = (state: RootState, props: any) => {
+    return {
+        buttonIsDisabled: state.download.downloads.findIndex((pub) => pub.url === props.publication.url) > -1,
```
Comment (danielweck) [inline src/renderer/components/dialog/publicationInfos/opdsControls.tsx]: I think `state.download.downloads.findIndex((pub) => pub.url === props.publication.url) > -1` can be simplified:
`state.download.downloads.find((pub) => pub.url === props.publication.url)`

Comment (danielweck) [reply src/renderer/components/dialog/publicationInfos/opdsControls.tsx]: Mmmh, maybe not :)
`buttonIsDisabled` is a boolean, so better not assign an object. Scrap this comment!

File: src/main/cli/commandLine.ts (hunk 8 above)
Comment (panaC) [reply src/main/cli/commandLine.ts]: no !

File: src/main/api/publication.ts (hunk 7 above)
Comment (panaC) [reply src/main/api/publication.ts]: I missed this.. I propose this fct : https://github.com/readium/readium-desktop/pull/793/commits/dcc68ce955bf2f5cab375784a5ed2667de908fd4

Comment (danielweck) [reply src/main/api/publication.ts]: Yes, until this:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled
:)

File: src/main/cli/commandLine.ts (hunk 8 above)
Comment (danielweck) [reply src/main/cli/commandLine.ts]: Sorry, it was a "doh" moment on my part!
(I misread the code inside the `if` block)

File: src/main/api/publication.ts (hunk 6 above)
Comment (danielweck) [reply src/main/api/publication.ts]: I think consistency is important here, so that all `dispatchToastRequest` calls are made from the same logical "block" in the app's architecture.

File: src/main/cli/commandLine.ts (hunk 8 above)
Comment (panaC) [reply src/main/cli/commandLine.ts]: In fact your are right, i may add an openReader loop on each publicationView. But it's not very useful on our case.

**Comment by panaC** [comment]
@danielweck i have to copy https://github.com/jasonrhodes/promise-all-settled/blob/master/index.js and paste into promise.ts or i add it in package.json ?

**Comment by danielweck** [comment]
Personally, I would favour the code snippet I posted in my first comment. I don't think we need the non-standard onlyFulfilled and onlyRejected API features.

File: src/main/api/publication.ts (hunk 6 above)
Comment (panaC) [reply src/main/api/publication.ts]: i don't know how to do this with the current nesting on our base code with `api` and `service`.
